grpc/binlog/v1alpha/binarylog.proto

```diff
@@ -61,6 +65,17 @@ message GrpcLogEntry {
   // Peer address information, will only be recorded in SEND_INITIAL_METADATA
   // and RECV_INITIAL_METADATA entries.
   Peer peer = 6;
+
+  // The method name. Always logged for every entry.
+  string full_method_name = 8;
+
+  // status_code and status_message:
+  // Only present for SEND_TRAILING_METADATA on server side or
+  // RECV_TRAILING_METADATA on client side.
+  uint32 status_code = 9;
+  // An original status message before any transport specific
+  // encoding.
+  string status_message = 10;
```

Comment:
Do we also want to log status details?

Reply:
Are there more details beyond status_code and status_message?

Reply:
Yes.  We can support arbitrary binary details, encoded as a `google.rpc.Status` proto, transmitted as the value of the `grpc-status-details-bin` metadata key.

Reply:
Although `grpc-status-details-bin` begins with the special `grpc-` prefix, it is something that is set at the application level via a utility. I think it's fine not to give it a top level field, and just treat it like a normal metadata entry because gRPC's public API does not know about it. But I do think that our tooling should make it easy to inspect the contents.

Reply:
@yang-g, in C++, don't we encode the error details in `grpc::Status` instead of returning it as metadata?

If so, it's really part of our API and the fact that we transmit it as metadata on the wire should be just as abstracted as the fact that we transmit the status code and message the same way.

Reply:
The error details does go into the Status object. But that is just for the C++ API level. The C API does not see the error details as special part of API like the error code and message. I agree it could be better to pluck it out and put it in a separate proto field, though I can see it is easier to just log it as part of metadata too.

Reply:
The C core API doesn't really matter; I think what matters is the C++ API, since that's what we expose to users.  (Also, I think the fact that the C-core API returns the error details in a different way is something we should fix.)

Reply:
@ejona86 @markdroth @dfawley PTAL

I audited Java and go and C:
- golang: [http2_client.go](https://github.com/grpc/grpc-go/blob/master/transport/http2_client.go#L356)
- java: [callers of Metadata.Key.of()](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/Metadata.java#L566)
- C: [gen_static_metadata.py](https://github.com/grpc/grpc/blob/master/tools/codegen/core/gen_static_metadata.py) 

binlog should exclude:
- anything that begins with "grpc-": these are set by gRPC and are not be needed. Exceptions: grpc-trace-bin and grpc-tags-bin are a gray area because it's not clear whether to treat census as an application, I can see these going either way. "grpc-status-details-bin" should be included.
- "lb-token": This token is handled by grpc. In fact, for java, this happens in the load balancer and is not visible to interceptors.
- {http2, speedy} transport specific headers: see list below
- call credentials

binlog should include:
- "grpc-status-details-bin": spoke with a11r and determined it's ok to log this as a metadata field, and make the tooling responsible for parsing it. The status code and message should already match the grpc status.
- stickiness service config key: since this is application controlled, and could very well be a pre-existing application key that is also used for load balancing. 
- "user-agent", ":authority" both are ambiguous but from past discussions it sounds like we want to log these.

Example of grpc- keys that should be excluded if visible:
"grpc-timeout" // log as top level proto field: timeout_ns
"grpc-status"  // log as top level proto field
"grpc-tags-bin"  // questionable
"grpc-trace-bin"  // questionable
"grpc-previous-rpc-attempts"
"grpc-retry-pushback-ms"
"grpc-encoding"
"grpc-accept-encoding"

Should be excluded if visible:
"lb-token": lb is lower level than what the application sees

http/2 metadata that should be excluded:
":path"  // log as top level proto field: method_name
":status"  // grpc status should be provided
":authority"
":method"
":scheme"
"content-encoding"
"accept-encoding"
"content-type"
"te"
"user-agent"
// spdy only
":host"
":version"
// gen_static_metadata.py http2 metadata
content-disposition
expires
... etc

Reply:
SGTM

Reply:
Btw please pay special consideration to excluding call credentials. These look easy to exclude for java and golang but I am not familiar enough with C to to tell.

Reply:
For the two ambiguous grpc- keys (grpc-trace-bin and grpc-tags-bin) I propose to exclude it, and have `grpc-status-details-bin` be the only exception to the rule.

Reply:
In general, this looks right to me.

I think the rule of thumb here should be "include only what would need to be specified by the application to replay the call".  That implies:

- Include `grpc-status-details-bin`, since that's specified by the application
- Include stickiness key, since that's specified by the application
- Exclude census metadata (will be added automatically to the replayed RPC by census)
- Exclude `user-agent` and `:authority` (will be automatically added to the replayed RPC by gRPC)

We should definitely exclude call credentials, but I don't think we can do this with a blanket rule, because call creds could use any arbitrary metadata key.  It will be up to individual implementations to make sure these are excluded.

Reply:
@markdroth Would it be worth adding some sort of "redacted" header to here?  Something that indicates that call creds were present, but have been removed?

Reply:
What would the use case be for that?

Reply:
Diagnotistic.  If  call fails due to bad creds, looking back in the logs would make it seem like the creds were never specified.  It would be similar to the truncated bit.

Reply:
If a call fails due to bad creds, it doesn't really matter whether the creds are missing or incorrect; either way, you will need to do further debugging to see what's actually being sent.  I'm not sure it really adds much to indicate this in the binary info logs.

Reply:
Not to belabor this, but can't that argument be made for *any* header then?

Reply:
Sure.  Which is why we also don't include an indication that any other header has been excluded from the logs.

Reply:
A recent CL has added a truncation flag for headers: https://github.com/grpc/grpc-proto/commit/22fb07542ca06f954bb6f384ffbb19f40be3232a , should this not be logged at all?

Reply:
No, I'm fine with that; it indicates that the log's payload data (whether it be metadata or headers) may not be complete.  I just don't see a need to indicate that each individual header we're excluding is being excluded.